xtask/src/commands/semver_check.rs

```diff
@@ -0,0 +1,132 @@
+    #[cfg(not(feature = "semver-checks"))]
+    {
+        let _ = workspace;
+        let _ = args;
+
+        return Err(anyhow::anyhow!(
+            "Feature `semver-checks` is not enabled. Use the `xcheck` alias",
+        ));
+    }
+
+    #[cfg(feature = "semver-checks")]
+    match args.command {
+        SemverCheckCmd::GenerateBaseline => {
+            checker::generate_baseline(&workspace, args.packages, args.chips)
+        }
+        SemverCheckCmd::Check => checker::check(&workspace, args.packages, args.chips),
+    }
+}
+
+#[cfg(feature = "semver-checks")]
+pub mod checker {
+    use std::{
+        fs,
+        io::Write,
+        path::{Path, PathBuf},
+    };
+
+    use cargo_semver_checks::ReleaseType;
+    use esp_metadata::Chip;
+
+    use crate::{
+        Package,
+        semver_check::{build_doc_json, minimum_update, remove_unstable_items},
+    };
+
+    pub fn generate_baseline(
+        workspace: &Path,
+        packages: Vec<Package>,
+        chips: Vec<Chip>,
+    ) -> Result<(), anyhow::Error> {
+        for package in packages {
+            log::info!("Generating API baseline for {package}");
+
+            for chip in &chips {
+                log::info!("Chip = {}", chip.to_string());
+                let package_name = package.to_string();
+                let package_path = crate::windows_safe_path(&workspace.join(&package_name));
+
+                let current_path = build_doc_json(package, chip, &package_path)?;
+
+                remove_unstable_items(&current_path)?;
+
+                let file_name = if package.chip_features_matter() {
+                    format!("{}", chip.to_string())
+                } else {
+                    "api".to_string()
+                };
+
+                let to_path = PathBuf::from(&package_path)
+                    .join(format!("api-baseline/{}.json.gz", file_name));
+                fs::create_dir_all(to_path.parent().unwrap())?;
+
+                log::debug!("Compress into {current_path:?}");
+                let mut encoder = flate2::write::GzEncoder::new(
+                    fs::File::create(to_path)?,
+                    flate2::Compression::default(),
+                );
+                encoder.write_all(&std::fs::read(current_path)?)?;
+
+                if !package.chip_features_matter() {
+                    break;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn check(
+        workspace: &Path,
+        packages: Vec<Package>,
+        chips: Vec<Chip>,
+    ) -> Result<(), anyhow::Error> {
+        for package in packages {
+            log::info!("Semver-check API for {package}");
+
+            for chip in &chips {
+                let result = minimum_update(workspace, package, *chip)?;
+                log::info!("Required bump = {:?}", result);
+
+                if result == ReleaseType::Major {
+                    return Err(anyhow::anyhow!("Semver check failed - needs a major bump"));
```

Comment:
This function is called `check` without any additional context, but why does it fail on a Major API change? Could you give this a better name so that we know what it does 3 months from now? `check_for_breaking_changes` would probably be enough?

Also I'd like this to not fail immediately, but collect the offending crates, and print all of them at once.